File: C/src/static_queue.c
#include <stdio.h>
#include "static_queue.h"

void inicializarStaticQueue(StaticQueue *q) {
    q->frente = 0;
    q->tras = -1;
    q->tamanho = 0;
}

int filaVazia(StaticQueue *q) {
    return q->tamanho == 0;
}

int filaCheia(StaticQueue *q) {
    return q->tamanho == MAXTAM;
}

int enfileirar(StaticQueue *q, Registro r) {
    if (filaCheia(q)) return 0;

    q->tras = (q->tras + 1) % MAXTAM;
    q->dados[q->tras] = r;
    q->tamanho++;
    return 1;
}

int desenfileirar(StaticQueue *q, Registro *r) {
    if (filaVazia(q)) return 0;

    *r = q->dados[q->frente];
    q->frente = (q->frente + 1) % MAXTAM;
    q->tamanho--;
    return 1;
}

void imprimirStaticQueue(StaticQueue *q) {
    if (filaVazia(q)) {
        printf("Fila vazia.\n");
        return;
    }

    int i = q->frente;
    for (int count = 0; count < q->tamanho; count++) {
        printf("%d ", q->dados[i].valor);
        i = (i + 1) % MAXTAM;
    }
    printf("\n");
}

File: Java/src/StaticQueue.java

public class StaticQueue {
    protected Registro[] elementos;
    protected int head;
    protected int tail;
    protected int size;
    protected int capacidade;

    public StaticQueue(int capacidade) {
        this.capacidade = capacidade;
        this.elementos = new Registro[capacidade];
        this.head = 0;
        this.tail = -1;
        this.size = 0;
    }

    public boolean enfileirar(Registro valor) {
        if (this.size == this.capacidade) return false; // Fila cheia

        this.tail = (this.tail + 1) % this.capacidade;
        this.elementos[this.tail] = valor;
        this.size++;
        return true;
    }

    public Registro desenfileirar() {
        if (filaVazia()) return null; // Fila vazia

        Registro valorRemovido = this.elementos[this.head];
        this.head = (this.head + 1) % this.capacidade;
        this.size--;
        return valorRemovido;
    }

    public boolean filaVazia() {
        return this.size == 0;
    }

    public void imprimir() {
        if (filaVazia()) {
            System.out.println("Fila vazia.");
            return;
        }
        for (int i = 0; i < this.size; i++) {
            System.out.print(this.elementos[(this.head + i) % this.capacidade].valor + " ");
        }
        System.out.println();
    }

    public void salvarEmArquivo(String nomeArquivo) {
        try (java.io.FileWriter writer = new java.io.FileWriter(nomeArquivo)) {
            for (int i = 0; i < this.size; i++) {
                writer.write(this.elementos[(this.head + i) % this.capacidade].valor + "\n");
            }
        } catch (java.io.IOException e) {
            System.err.println("Erro ao abrir arquivo para escrita: " + e.getMessage());
        }
    }
}

